elastic/kibana
# [Upgrade Assistant] Remove "boom" from reindex service

## Summary

This contribution replaces the error handling inside reindex service with custom error handling (which was prone to errors could either be Boom HTTP errors or actual Errors which led to complications in handling). Custom Errors are also platform prescribed: https://github.com/elastic/kibana/issues/57458. The mapping to HTTP specific codes is moved to just the handlers (see `mapAnyErrorToKibanaHttpResponse` inside `reindex_indices`).

## How to test

The existing Jest test coverage tests that all throws are still occurring when expected.

This was integration tested locally using this setup:

1. Paste the code snippet into `cluster_checkup.ts` over the handler
2. Inside the reindex_service, remove the `if` statement around the `throw` statement on line 287. This will demonstrate an integration test.

<details>
<summary>Hack cluster_checkup.ts, ln 32</summary>

```ts

versionCheckHandlerWrapper(async (
        {
          core: {
            elasticsearch: { dataClient },
          },
        },
        request,
        response
      )
 => {
      // post migration
      const body = await getUpgradeAssistantStatus(
        dataClient,
        isCloudEnabled,
      );

      body.indices = body.indices.concat([{
        level: 'none',
        message: 'none level test',
        url: 'test/url',
        details: 'test details',
        index: 'test1',
        reindex: true,
        needsDefaultFields: false,
      }, {
        level: 'info',
        message: 'info level test',
        url: 'test/url',
        details: 'test details',
        index: 'test2',
        reindex: true,
        needsDefaultFields: false,
      }, {
        level: 'warning',
        message: 'warning level test2',
        url: 'test/url',
        details: 'test details',
        index: 'test3',
        reindex: true,
        needsDefaultFields: false,
      }, {
        level: 'critical',
        message: 'metricbeat-1 successful fix',
        url: 'test/url',
        details: 'test details',
        index: 'metricbeat-1',
        reindex: false,
        needsDefaultFields: true,
      }, {
        level: 'critical',
        message: 'metricbeat-2 failed fix',
        url: 'test/url',
        details: 'test details',
        index: 'metricbeat-2',
        reindex: false,
        needsDefaultFields: true,
      }]);


      body.cluster = body.cluster.concat([{
        level: 'none',
        message: 'none level test message',
        url: 'test/url',
        details: 'test details',
      }, {
        level: 'info',
        message: 'info level test message',
        url: 'test/url',
        details: 'test details',
      }, {
        level: 'warning',
        message: 'warning level test message',
        url: 'test/url',
        details: 'test details',
      }, {
        level: 'critical',
        message: 'critical level test message',
        url: 'test/url',
        details: 'test details',
      }]);

      try {
        return response.ok({
          body,
        });
      } catch (e) {
        if (e.status === 403) {
          return response.forbidden(e.message);
        }

        return response.internalError({ body: e });
      }
    })
```
</details>

## To reviewers

1. Currently, reindex service is only used in the reindex routes and inside of reindex worker. Specifically, the `processNextStep` function is used inside of reindex worker. These cases are all covered and tested, flag it here if you can find any instance of reindex service being used (and errors being caught) that I may have missed.
2. I am using `MyError extends Error` with `symbol` to uniquely identify errors. Open to discussing this approach here too.

### Additional

* Currently there is only one place that a boom check is in server side UA - reindex actions. This is an `isBoom` check against an error that could be emitted from Saved Objects client.

### For maintainers
